File: csharp/test/TEST_IpcConnection.cs
using System;
using System.Collections;
using System.Collections.Generic;

using Xunit;

using InoIPC;

namespace InoIPC.TEST
{
   // ============================================================
   /// <summary>
   /// Tests for IpcConnection: send, receive, request pattern.
   /// </summary>
   // ============================================================
   public class TEST_IpcConnection
   {

   #region Send / Receive

      [Fact]
      public void Send_Receive_RoundTrip()
      {
         var transport = new TestTransport();
         var conn      = new IpcConnection(transport);

         conn.Send("{\"test\":\"ping\"}");

         string result = conn.Receive();

         Assert.Equal("{\"test\":\"ping\"}", result);
      }

      [Fact]
      public void Send_IpcResponse()
      {
         var transport = new TestTransport();
         var conn      = new IpcConnection(transport);

         conn.Send(IpcResponse.Success("ok"));

         string result = conn.Receive();

         Assert.Contains("\"success\":true", result);
         Assert.Contains("\"message\":\"ok\"", result);
      }

   #endregion

   #region Request

      [Fact]
      public void Request_ParsesResponse()
      {
         var transport = new TestTransport();

         // Simulate server response in buffer
         FrameProtocol.Send(transport, "{\"success\":true,\"value\":42}");

         var conn     = new IpcConnection(transport);
         var response = conn.Request("{\"action\":\"get\"}");

         Assert.True(response.IsSuccess);
      }

   #endregion

   #region Server Integration

      [Fact]
      public void Server_Sends_Client_Receives()
      {
         var transport = new TestTransport();

         // Server writes
         var serverConn = new IpcConnection(transport);
         serverConn.Send(IpcResponse.Success("ready"));

         // Client reads from same buffer
         var clientConn = new IpcConnection(transport);
         string response = clientConn.Receive();

         Assert.Contains("\"success\":true", response);
         Assert.Contains("\"message\":\"ready\"", response);
      }

      [Fact]
      public void TestServer_Callback_Receives_IpcConnection()
      {
         var server   = new TestServer();
         bool called  = false;

         server.Start(conn =>
         {
            called = true;
            Assert.NotNull(conn);
         });

         server.Accept();

         Assert.True(called);
      }

   #endregion

   }
}

File: rust/inoipc/tests/test_ipc_connection.rs
use inoipc::{IpcConnection, IpcResponse, TestServer, TestTransport, frame_send};

// ----------------------------------------------------------------
// Send / Receive
// ----------------------------------------------------------------

#[test]
fn send_receive_round_trip() {
    let transport = TestTransport::new();
    let mut conn = IpcConnection::new(transport);
    conn.send("{\"test\":\"ping\"}").unwrap();
    let result = conn.receive().unwrap();
    assert_eq!(result, "{\"test\":\"ping\"}");
}

#[test]
fn send_ipc_response() {
    let transport = TestTransport::new();
    let mut conn = IpcConnection::new(transport);
    conn.send_response(&IpcResponse::success_msg("ok")).unwrap();
    let result = conn.receive().unwrap();
    assert!(result.contains("\"success\":true"));
    assert!(result.contains("\"message\":\"ok\""));
}

// ----------------------------------------------------------------
// Request
// ----------------------------------------------------------------

#[test]
fn request_parses_response() {
    let mut transport = TestTransport::new();
    // Pre-load server response into the buffer.
    frame_send(&mut transport, "{\"success\":true,\"value\":42}").unwrap();
    let mut conn = IpcConnection::new(transport);
    let response = conn.request("{\"action\":\"get\"}").unwrap();
    assert!(response.is_success());
}

// ----------------------------------------------------------------
// Server integration
// ----------------------------------------------------------------

#[test]
fn server_sends_client_receives() {
    let transport = TestTransport::new();
    let mut server_conn = IpcConnection::new(transport.clone());
    server_conn.send_response(&IpcResponse::success_msg("ready")).unwrap();

    let mut client_conn = IpcConnection::new(transport);
    let response = client_conn.receive().unwrap();
    assert!(response.contains("\"success\":true"));
    assert!(response.contains("\"message\":\"ready\""));
}

#[test]
fn test_server_callback_receives_connection() {
    let mut server = TestServer::new();
    let called = std::sync::Arc::new(std::sync::atomic::AtomicBool::new(false));
    let called_clone = called.clone();

    server.start(move |_conn| {
        called_clone.store(true, std::sync::atomic::Ordering::Relaxed);
    });

    server.accept().unwrap();
    assert!(called.load(std::sync::atomic::Ordering::Relaxed));
}
